Donation-platform/src/main/java/com/google/FindTrendingCharities.java

```diff
@@ -30,144 +30,152 @@
 
 public final class FindTrendingCharities {
 
-  private DbCalls db;
-  private DatastoreService ds;
-
-  // number of trending charities to be returned
-  final int MAX_NUM_OF_CHARITIES_TO_RETURN = 7;
-
-  final double CHARITY_NAV_SCALE_FACTOR = 1.25;
-
-  // should sum to 1
-  final double USER_RATING_WEIGHT = 0.75;
-  final double CHARITY_NAV_WEIGHT = 0.25;
-
-  // should sum to 1
-  final double TAGS_SCORE_WEIGHT = 0.75;
-  final double AVG_REVIEW_WEIGHT = 0.25;
-
-  private static Collection<Charity> charities;
-
-  //constructor to do set up
-  public FindTrendingCharities(DatastoreService ds) {
-    //ds = DatastoreServiceFactory.getDatastoreService();
-    this.ds = ds;
-    db = new DbCalls(ds);
-    DbSetUpUtils setUp = new DbSetUpUtils(ds, db);
-    this.charities = getAllCharities();
-    // only populate database if there is nothing in the database already
-    if (charities.size() < 1) {
-        setUp.populateDatabase();
-    }
-  }
-
-  // returns the collection of top trending charities
-  public Collection<Charity> queryDb() {
-    //Collection<Charity> charities = getAllCharities();
-    for (Charity charity : charities) {
-      double charityScore = calcCharityTrendingScore(charity);
-      charity.setTrendingScoreCharity(charityScore);
-      try {
-          db.updateCharity(charity);
-      }
-      catch (Exception e) {
-          System.out.println("unable to update charity: " + e);
-      }
+    private DbCalls db;
+    private DatastoreService ds;
+
+    // number of trending charities to be returned
+    final int MAX_NUM_OF_CHARITIES_TO_RETURN = 7;
+
+    final double CHARITY_NAV_SCALE_FACTOR = 1.25;
+
+    // should sum to 1
+    final double USER_RATING_WEIGHT = 0.75;
+    final double CHARITY_NAV_WEIGHT = 0.25;
+
+    // should sum to 1
+    final double TAGS_SCORE_WEIGHT = 0.75;
+    final double AVG_REVIEW_WEIGHT = 0.25;
+
+    private static Collection<Charity> charities;
+    private List<Charity> topTrendingCharities;
+
+    //constructor to do set up
+    public FindTrendingCharities(DatastoreService ds) {
+        //ds = DatastoreServiceFactory.getDatastoreService();
+        this.ds = ds;
+        db = new DbCalls(ds);
+        DbSetUpUtils setUp = new DbSetUpUtils(ds, db);
+        if (charities == null) {
+            this.charities = getAllCharities();
+            if (charities.size() < 1) {
+                setUp.populateDatabase();
```

Comment:
why are we making this check here. Shouldn't our db always be populated. if not, this is going to add latency to populatecharities. this function should only do read call and if db is empty we should just fail gracefully.

Reply:
The db is always populated, but for the first time FindTrendingCharities is instantiated from anywhere, we need to getAllCharities() from the db into memory (this is the charities class variable) to process them.

Reply:
getAllCharities is ok, the question is why are we checking on size and populating then. Probably just check if its is non-null and throw error otherwise instead of populating. This all is otherwise going to add un-necessary latency

Reply:
ok.